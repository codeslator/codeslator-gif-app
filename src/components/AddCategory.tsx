import { useState, FormEvent, SyntheticEvent, FC } from 'react';
import { AddCategoryProps } from '../interfaces/index';

export const AddCategory: FC<AddCategoryProps> = ({ categories, setCategories }) => {
  const [value, setValue] = useState<string>('');

  const handleInput = (event: FormEvent<HTMLInputElement>) => {
    setValue(event.currentTarget.value)
  };

  const handleSubmit = (e: SyntheticEvent) => {
    e.preventDefault();
    if(value !== '') {
      setCategories((prev: string[]) => [...prev, value]);
      setValue('');
    }
  };

  return (
    <>
      <form onSubmit={handleSubmit}>
        <h2>Search Category</h2>
        <div className="row">
          <div className="col-9">
            <input
              type="text"
              className="form-control"
              placeholder="Category"
              id="category"
              value={value}
              onChange={handleInput}
            />
          </div>
          <div className="col-3">
            <button className="btn btn-primary btn-full" type="submit">Submit</button>
          </div>
        </div>
      </form>
      <p>{value}</p>
    </>
  )
}
