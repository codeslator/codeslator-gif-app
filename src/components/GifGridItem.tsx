import { GifGridItemProps } from '../interfaces/index';

export const GifGridItem = ({ gif }: GifGridItemProps) => {
  const { title, url } = gif;
  return (
    <div className="card mb-3 animate__animated animate__fadeIn">
      <img src={url} className="card-img-top" alt={title} />
      <div className="card-body">
        <h5 className="card-title">{title}</h5>
      </div>
    </div>
  )
}
