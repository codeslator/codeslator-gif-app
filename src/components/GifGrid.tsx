import { FC, useEffect } from 'react';
import { GifGridProps } from '../interfaces/index';
import useGif from '../hooks/useGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid: FC<GifGridProps> = ({ category }) => {
  const { fetchGifs, gifs, isLoading } = useGif();

  useEffect(() => {
    fetchGifs(category)
  }, [category, fetchGifs])

  return (
    <>
      <h3 className="mb-3">{category}</h3>
      <div className="row">
        {isLoading && (
          <h4 className="text-center">Loading...</h4>
        )}
        {gifs.map((gif) => (
          <div className="col-xs-6 col-sm-4 col-md-3 col-lg-2" key={gif.id} >
            <GifGridItem gif={gif} />
          </div>
        ))}
      </div>
    </>
  )
}
