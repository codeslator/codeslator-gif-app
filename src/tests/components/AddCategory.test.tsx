import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import { AddCategory } from '../../components/AddCategory';

describe('AddCategory Component', () => {
  const setCategories = jest.fn(); // Mocking the function
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />)
  })
  
  test('Should to show the component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should to change value in the input', () => {
    const input = wrapper.find('input');
    const value = 'Hello world';
    input.simulate('change', { currentTarget: { value } }); // Simulate an event
    const p = wrapper.find('p');
    expect(p.text().trim()).toBe(value)
  });

  test('Don\'t should submit the data', () => {
    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
    expect(setCategories).not.toHaveBeenCalled();
  });

  test('Should submit data, call setCategories and clear the input', () => {
    const value = 'Hello world';
    wrapper.find('input').simulate('change', { currentTarget: { value } }); // Simulate an event
    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
    expect(wrapper.find('input').prop('value')).toBe('')
  });
})