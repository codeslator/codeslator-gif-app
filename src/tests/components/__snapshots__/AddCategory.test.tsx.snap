// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory Component Should to show the component correctly 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Search Category
    </h2>
    <div
      className="row"
    >
      <div
        className="col-9"
      >
        <input
          className="form-control"
          id="category"
          onChange={[Function]}
          placeholder="Category"
          type="text"
          value=""
        />
      </div>
      <div
        className="col-3"
      >
        <button
          className="btn btn-primary btn-full"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
  <p />
</Fragment>
`;
